In [1]:
import pandas as pd
import numpy as np
from datetime import datetime, timedelta
import pickle
import requests

from bokeh.io import show
from bokeh.models import ColumnDataSource, GMapOptions, HoverTool, GlyphRenderer, Legend
from bokeh.plotting import gmap, figure
from bokeh.layouts import column

from IPython.core.display import display, HTML
display(HTML("<style>.container { width:95% !important; }</style>"))

# Oslo Bysykkel scheme quicklook

A quick look at load balancing in Oslo based on a day worth of station capacity data collected from their api.

I ran the counterpart script `collect_data.py` every 10 minutes for a 24 hour period in order to look for which stations were emptied and refilled.

In [2]:
secrets = dict()
with open('secrets', 'r') as f:
    for line in f:
        l = line.strip('\n').split('=')
        secrets[l[0]] = l[1]
secrets.keys()

dict_keys(['GMAPS_API_KEY', 'OSLO_BYSYKKEL_API_KEY'])

In [3]:
base_url = 'https://oslobysykkel.no/api/v1'
auth_header = {'Client-Identifier': secrets['OSLO_BYSYKKEL_API_KEY']}

r = requests.get(f'{base_url}/stations', headers=auth_header)

stations = r.json()['stations']

In [4]:
capacity = dict()
for s in stations:
    capacity[s['id']] = s['number_of_locks']
    
station_names = dict()
for s in stations:
    station_names[s['id']] = s['title']
    
station_locs = dict()
for s in stations:
    station_locs[s['id']] = s['center']

Get the records and format them as needed. This means attaching the station details to each record and correcting for timezone.

In [5]:
df = pd.read_csv('oslo/oslo_sykkel_records.csv', index_col=0)
df = df.drop(['overflow_capacity'], axis=1)
df['date'] = pd.to_datetime(df['date']) + timedelta(seconds = 60*60*2)
df['lat'] = df['id'].apply(lambda x: station_locs.get(x, {'latitude': np.nan})['latitude'])
df['lon'] = df['id'].apply(lambda x: station_locs.get(x, {'longitude': np.nan})['longitude'])
df['name'] = df['id'].map(station_names)
df['capacity'] = df['id'].map(capacity)

ids = set(df['id'])
print(len(ids))
df.head()

299


bikes   id  locks                date        lat        lon  \
0      0  177     28 2018-09-04 12:49:52  59.915629  10.777664   
1     15  166      4 2018-09-04 12:49:52  59.943748  10.760195   
2      0  178     12 2018-09-04 12:49:52  59.929838  10.711223   
3      1  167     16 2018-09-04 12:49:52  59.920565  10.734274   
4      1  173      6 2018-09-04 12:49:52  59.916847  10.758148   

                          name  capacity  
0                  Tøyenparken      29.0  
1            Bjølsen Studentby      21.0  
2               Colosseum Kino      12.0  
3  Pilestredet Park Studenthus      18.0  
4                   Legevakten      10.0

Capacity seems to shift. bikes + locks changes across the day for a single station, eg id=177 goes from 28 to 24 to 25. Sometimes will hit maximum gained from the /stations endpoint.

### Looking for refills

We can look for the change in number of bikes between each record and use that to guess if there was a refill.

There's a chance of course that finer measurements are needed and I should have used every 2 or 5 minutes instead of 10 but let's see.

In [6]:
df = df.sort_values(by=['id', 'date'])

df['diff'] = df['bikes'] - df['bikes'].shift(1)

earliest_date = df['date'].min()
df = df[df['date'] != earliest_date].reset_index(drop=True)

df.head()

bikes   id  locks                date       lat        lon          name  \
0      4  157     26 2018-09-04 13:00:01  59.91562  10.762248  Nylandsveien   
1      4  157     26 2018-09-04 13:09:58  59.91562  10.762248  Nylandsveien   
2      3  157     27 2018-09-04 13:19:55  59.91562  10.762248  Nylandsveien   
3      4  157     26 2018-09-04 13:29:51  59.91562  10.762248  Nylandsveien   
4      4  157     26 2018-09-04 13:40:01  59.91562  10.762248  Nylandsveien   

   capacity  diff  
0      30.0   0.0  
1      30.0   0.0  
2      30.0  -1.0  
3      30.0   1.0  
4      30.0   0.0

This is a reasonably busy station in the centre of town. The -22 and the late-night -10 could well be load-balancing. We can use something like +-8 for our cutoff for now.

In [7]:
df[(df['id'] == 389) & (abs(df['diff']) >= 4)]

bikes   id  locks                date        lat        lon  \
32955     37  389      1 2018-09-04 14:49:57  59.911881  10.747759   
32966     15  389     22 2018-09-04 16:39:58  59.911881  10.747759   
32979     10  389     26 2018-09-04 18:50:00  59.911881  10.747759   
33011      6  389     31 2018-09-05 00:09:52  59.911881  10.747759   
33025      2  389     35 2018-09-05 02:29:59  59.911881  10.747759   
33059      8  389     30 2018-09-05 08:09:57  59.911881  10.747759   
33065     19  389     19 2018-09-05 09:09:51  59.911881  10.747759   
33078     22  389     16 2018-09-05 11:19:52  59.911881  10.747759   

              name  capacity  diff  
32955  Arkaden Sør      39.0   5.0  
32966  Arkaden Sør      39.0 -22.0  
32979  Arkaden Sør      39.0  -4.0  
33011  Arkaden Sør      39.0  -7.0  
33025  Arkaden Sør      39.0  -4.0  
33059  Arkaden Sør      39.0   4.0  
33065  Arkaden Sør      39.0   4.0  
33078  Arkaden Sør      39.0 -10.0

In [8]:
df.sort_values(by="diff")

bikes   id  locks                date        lat        lon  \
32966     15  389     22 2018-09-04 16:39:58  59.911881  10.747759   
39214     17  465     22 2018-09-04 16:39:58  59.910820  10.730321   
9294      15  222     19 2018-09-04 23:40:01  59.922414  10.758127   
97        10  157     20 2018-09-05 05:09:55  59.915620  10.762248   
31601      0  379     20 2018-09-05 01:49:59  59.914837  10.753622   
452        7  160     42 2018-09-04 17:19:59  59.910215  10.751687   
9261      14  222     19 2018-09-04 18:09:58  59.922414  10.758127   
12780     14  247     30 2018-09-04 13:00:01  59.910091  10.752138   
11075     23  234     18 2018-09-05 12:30:00  59.913731  10.736032   
28752     10  359     18 2018-09-05 00:19:56  59.911609  10.761874   
9266      14  222     20 2018-09-04 19:00:02  59.922414  10.758127   
19695      1  295     18 2018-09-05 05:29:52  59.915412  10.758343   
6691      21  204     27 2018-09-04 15:49:57  59.910287  10.750012   
19992     12  297     12 2018-09-05 07:39:52  59.909056  10.756217   
19888      3  297     27 2018-09-04 14:19:55  59.909056  10.756217   
12910      4  247     40 2018-09-05 10:39:52  59.910091  10.752138   
26504      4  343     22 2018-09-05 04:20:00  59.917258  10.758114   
37271      0  419     28 2018-09-05 00:09:52  59.912193  10.748067   
39863      3  469     18 2018-09-05 06:09:58  59.914591  10.759867   
17440     19  279     15 2018-09-05 08:19:56  59.908353  10.741794   
934        0  163     39 2018-09-05 02:39:54  59.913485  10.757406   
40447      1  473     27 2018-09-05 08:49:52  59.924354  10.761508   
12278      7  243     15 2018-09-04 23:59:56  59.915471  10.751096   
16658      6  274     25 2018-09-04 20:19:57  59.905031  10.753658   
12809     12  247     30 2018-09-04 17:49:58  59.910091  10.752138   
14663     24  260     32 2018-09-04 19:09:53  59.910330  10.705145   
12216     12  243     12 2018-09-04 13:40:01  59.915471  10.751096   
19336      7  293     24 2018-09-04 17:00:01  59.920352  10.683687   
442       20  160     28 2018-09-04 15:40:01  59.910215  10.751687   
40441      8  473     21 2018-09-05 07:49:53  59.924354  10.761508   
...      ...  ...    ...                 ...        ...        ...   
1886      14  170      4 2018-09-04 19:40:01  59.930596  10.714796   
3338      14  180      6 2018-09-05 00:59:53  59.930912  10.791843   
35865     21  409      5 2018-09-05 02:29:59  59.947497  10.732429   
35600     14  407     15 2018-09-05 05:40:02  59.941148  10.750710   
705       12  161      7 2018-09-05 11:49:58  59.926872  10.738456   
21055     19  305     22 2018-09-04 19:29:53  59.938998  10.723075   
17438     25  279     10 2018-09-05 07:59:54  59.908353  10.741794   
3555      13  182      0 2018-09-04 13:49:52  59.937737  10.751701   
40446     14  473     15 2018-09-05 08:39:57  59.924354  10.761508   
36496     19  414      7 2018-09-04 13:19:55  59.929468  10.769051   
25912     14  339     13 2018-09-05 00:19:56  59.913116  10.789860   
41251     20  479     20 2018-09-05 00:49:59  59.929347  10.713739   
35712     15  408     16 2018-09-05 00:40:00  59.945417  10.713804   
15034     15  262     15 2018-09-05 09:39:56  59.923159  10.766055   
11710     15  239     11 2018-09-04 23:59:56  59.924335  10.792799   
15644     17  267      5 2018-09-04 17:00:01  59.925480  10.731318   
2067      15  171      9 2018-09-05 02:09:57  59.940380  10.730068   
3778      21  183     17 2018-09-05 03:19:59  59.946699  10.773978   
34033     17  396      2 2018-09-05 04:49:58  59.952695  10.751254   
35671     16  408     15 2018-09-04 17:49:58  59.945417  10.713804   
3717      16  183     21 2018-09-04 17:09:59  59.946699  10.773978   
824       19  162      8 2018-09-05 07:59:54  59.934934  10.749475   
838       19  162      9 2018-09-05 10:19:48  59.934934  10.749475   
21134     19  305     22 2018-09-05 08:39:57  59.938998  10.723075   
41028     18  477      2 2018-09-05 10:59:52  59.930645  10.746299   
10561     22  231

In [9]:
map_options = GMapOptions(lat=59.914, lng=10.751, map_type="roadmap", zoom=11)

df_added = df[df['diff'] > 8]
df_removed = df[df['diff'] < -8]

added_data = dict(
    lat=list(df_added['lat'].values),
    lon=list(df_added['lon'].values),
)
removed_data = dict(
    lat=list(df_removed['lat'].values),
    lon=list(df_removed['lon'].values),
)

added_source = ColumnDataSource(added_data)
removed_source = ColumnDataSource(removed_data)


plot = gmap(secrets['GMAPS_API_KEY'], map_options, title="Oslo")

plot.circle(x="lon", y="lat", size=15, fill_color="blue", fill_alpha=0.8, source=added_source)
plot.circle(x="lon", y="lat", size=15, fill_color="red", fill_alpha=0.8, source=removed_source)


dumx = range(10)
dumfig = figure(outline_line_alpha=0, plot_height=200,
                plot_width=300, toolbar_location=None)
dumfig.circle(dumx, dumx, color='blue',
              size=10, legend='Bikes added')
dumfig.circle(dumx, dumx, color='red',
              size=10, legend='Bikes removed')

dumfig.renderers = [rend for rend in dumfig.renderers if (
    type(rend) == GlyphRenderer or type(rend) == Legend)]
dumfig.renderers[0].border_line_alpha = 1
dumfig.renderers[0].spacing = 6
dumfig.renderers[0].location = 'top_left'
for rend in dumfig.renderers:
    if type(rend) == GlyphRenderer:
        rend.visible = False
        
show(column(plot, dumfig))

Produces a live version of:

![title](oslo/oslo_load_balancing.png)